MC leader controller will crash when ClusterSet is deleted

**Describe the bug**
When the ClusterSet is deleted in a Namespace where the leader controller is running, it will raise following nil pointer error which causes the leader controller crash and restart.

```
leader_clusterset_controller.go:72] "Received ClusterSet delete" config=""
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x188b9b1]

goroutine 274 [running]:
antrea.io/antrea/multicluster/controllers/multicluster.(*MemberClusterAnnounceReconciler).processMCSStatus(0xc00052ff20)
	/antrea/multicluster/controllers/multicluster/memberclusterannounce_controller.go:181 +0x331
...
```

**To Reproduce**
Create a ClusterSet on the leader controller's Namespace, then delete it.

**Expected**
MC controller should be able to handle ClusterSet deletion event successfully.

**Additional context**
Besides fixing this bug, we need to add more e2e test cases to verify this kind of issue.

--- a/multicluster/controllers/multicluster/leader_clusterset_controller.go
+++ b/multicluster/controllers/multicluster/leader_clusterset_controller.go
@@ -69,7 +69,7 @@ func (r *LeaderClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		if !errors.IsNotFound(err) {
 			return ctrl.Result{}, err
 		}
-		klog.InfoS("Received ClusterSet delete", "config", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet delete", "clusterset", klog.KObj(clusterSet))
 		for _, removedMember := range r.clusterSetConfig.Spec.Members {
 			r.StatusManager.RemoveMember(common.ClusterID(removedMember.ClusterID))
 		}
@@ -80,7 +80,7 @@ func (r *LeaderClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{}, nil
 	}
 
-	klog.InfoS("Received ClusterSet add/update", "config", klog.KObj(clusterSet))
+	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
 	// Handle create or update
 	// If create, make sure the local ClusterClaim is part of the leader config
@@ -175,8 +175,8 @@ func (r *LeaderClusterSetReconciler) runBackgroundTasks() {
 //       statues across all clusters. Message will be empty and Reason
 //       will be "NoReadyCluster"
 func (r *LeaderClusterSetReconciler) updateStatus() {
-	defer r.mutex.Unlock()
 	r.mutex.Lock()
+	defer r.mutex.Unlock()
 
 	if r.clusterSetConfig == nil {
 		// Nothing to do.

--- a/multicluster/controllers/multicluster/member_clusterset_controller.go
+++ b/multicluster/controllers/multicluster/member_clusterset_controller.go
@@ -86,7 +86,7 @@ func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		if !apierrors.IsNotFound(err) {
 			return ctrl.Result{}, err
 		}
-		klog.InfoS("Received ClusterSet delete", "config", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet delete", "clusterset", klog.KObj(clusterSet))
 		stopErr := r.remoteCommonAreaManager.Stop()
 		r.remoteCommonAreaManager = nil
 		r.clusterSetConfig = nil
@@ -96,7 +96,7 @@ func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{}, stopErr
 	}
 
-	klog.InfoS("Received ClusterSet add/update", "config", klog.KObj(clusterSet))
+	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
 	// Handle create or update
 	if r.clusterSetConfig == nil {

--- a/multicluster/controllers/multicluster/memberclusterannounce_controller.go
+++ b/multicluster/controllers/multicluster/memberclusterannounce_controller.go
@@ -72,8 +72,8 @@ type MemberClusterAnnounceReconciler struct {
 }
 
 type MemberClusterStatusManager interface {
-	AddMember(MemberId common.ClusterID)
-	RemoveMember(MemberId common.ClusterID)
+	AddMember(memberID common.ClusterID)
+	RemoveMember(memberID common.ClusterID)
 
 	GetMemberClusterStatuses() []multiclusterv1alpha1.ClusterStatus
 }
@@ -106,8 +106,8 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	defer r.mapLock.Unlock()
 	r.mapLock.Lock()
+	defer r.mapLock.Unlock()
 
 	if data, ok := r.timerData[common.ClusterID(memberAnnounce.ClusterID)]; ok {
 		klog.V(2).InfoS("Reset lastUpdateTime", "Cluster", memberAnnounce.ClusterID)
@@ -163,8 +163,8 @@ func (r *MemberClusterAnnounceReconciler) SetupWithManager(mgr ctrl.Manager) err
 }
 
 func (r *MemberClusterAnnounceReconciler) processMCSStatus() {
-	defer r.mapLock.Unlock()
 	r.mapLock.Lock()
+	defer r.mapLock.Unlock()
 
 	for member, data := range r.timerData {
 		if data.lastUpdateTime.IsZero() {
@@ -172,11 +172,11 @@ func (r *MemberClusterAnnounceReconciler) processMCSStatus() {
 			continue
 		}
 		status := r.memberStatus[member]
-		// Check if the member has connected atleast once in the last 3 intervals.
+		// Check if the member has connected at least once in the last 3 intervals.
 		duration := time.Since(data.lastUpdateTime)
 		klog.V(2).InfoS("Timer processing", "Cluster", member, "duration", duration)
 		if duration <= ConnectionTimeout {
-			// Member has updated MemberClusterStatus atleast once in the last 3 intervals.
+			// Member has updated MemberClusterStatus at least once in the last 3 intervals.
 			// If last status is not connected, then update the status.
 			for index := range status.Conditions {
 				condition := &status.Conditions[index]
@@ -232,10 +232,10 @@ func (r *MemberClusterAnnounceReconciler) processMCSStatus() {
 
 /******************************* MemberClusterStatusManager methods *******************************/
 
-func (r *MemberClusterAnnounceReconciler) AddMember(MemberId common.ClusterID) {
-	defer r.mapLock.Unlock()
+func (r *MemberClusterAnnounceReconciler) AddMember(memberID common.ClusterID) {
 	r.mapLock.Lock()
-	if _, ok := r.memberStatus[MemberId]; ok {
+	defer r.mapLock.Unlock()
+	if _, ok := r.memberStatus[memberID]; ok {
 		// already present
 		return
 	}
@@ -256,30 +256,25 @@ func (r *MemberClusterAnnounceReconciler) AddMember(MemberId common.ClusterID) {
 		Reason:             ReasonNeverConnected,
 	})
 
-	r.memberStatus[MemberId] = &multiclusterv1alpha1.ClusterStatus{ClusterID: string(MemberId),
+	r.memberStatus[memberID] = &multiclusterv1alpha1.ClusterStatus{ClusterID: string(memberID),
 		Conditions: conditions}
 
-	r.timerData[MemberId] = &timerData{connected: false, lastUpdateTime: time.Time{}}
+	r.timerData[memberID] = &timerData{connected: false, lastUpdateTime: time.Time{}}
+	klog.InfoS("Added member", "member", memberID)
 }
 
-func (r *MemberClusterAnnounceReconciler) RemoveMember(MemberId common.ClusterID) {
-	defer r.mapLock.Unlock()
+func (r *MemberClusterAnnounceReconciler) RemoveMember(memberID common.ClusterID) {
 	r.mapLock.Lock()
+	defer r.mapLock.Unlock()
 
-	if _, ok := r.memberStatus[MemberId]; ok {
-		delete(r.memberStatus, MemberId)
-		return
-	}
-
-	if _, ok := r.timerData[MemberId]; ok {
-		delete(r.timerData, MemberId)
-		return
-	}
+	delete(r.memberStatus, memberID)
+	delete(r.timerData, memberID)
+	klog.InfoS("Removed member", "member", memberID)
 }
 
 func (r *MemberClusterAnnounceReconciler) GetMemberClusterStatuses() []multiclusterv1alpha1.ClusterStatus {
-	defer r.mapLock.Unlock()
 	r.mapLock.Lock()
+	defer r.mapLock.Unlock()
 
 	status := make([]multiclusterv1alpha1.ClusterStatus, len(r.memberStatus))
 

--- a/multicluster/controllers/multicluster/memberclusterannounce_controller_test.go
+++ b/multicluster/controllers/multicluster/memberclusterannounce_controller_test.go
@@ -114,10 +114,22 @@ func TestStatusAfterAdd(t *testing.T) {
 		},
 	}
 
-	//[]multiclusterv1alpha1.ClusterStatus
-	status := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
-	assert.Equal(t, 1, len(status))
-	verifyStatus(t, expectedStatus, status[0])
+	actualStatus := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
+	actualTimerData := memberClusterAnnounceReconcilerUnderTest.timerData
+	assert.Equal(t, 1, len(actualStatus))
+	assert.Equal(t, 1, len(actualTimerData))
+	verifyStatus(t, expectedStatus, actualStatus[0])
+}
+
+func TestStatusAfterDelete(t *testing.T) {
+	setup()
+	memberClusterAnnounceReconcilerUnderTest.AddMember("east")
+	memberClusterAnnounceReconcilerUnderTest.RemoveMember("east")
+
+	actualStatus := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
+	actualTimerData := memberClusterAnnounceReconcilerUnderTest.timerData
+	assert.Equal(t, 0, len(actualStatus))
+	assert.Equal(t, 0, len(actualTimerData))
 }
 
 func TestStatusAfterReconcile(t *testing.T) {
@@ -159,10 +171,10 @@ func TestStatusAfterReconcile(t *testing.T) {
 	}
 
 	memberClusterAnnounceReconcilerUnderTest.processMCSStatus()
-	status := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
-	klog.V(2).InfoS("Received", "actual", status, "expected", expectedStatus)
-	assert.Equal(t, 1, len(status))
-	verifyStatus(t, expectedStatus, status[0])
+	actualStatus := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
+	klog.V(2).InfoS("Received", "actual", actualStatus, "expected", expectedStatus)
+	assert.Equal(t, 1, len(actualStatus))
+	verifyStatus(t, expectedStatus, actualStatus[0])
 }
 
 func TestStatusAfterLeaderElection(t *testing.T) {
@@ -201,10 +213,10 @@ func TestStatusAfterLeaderElection(t *testing.T) {
 	}
 
 	memberClusterAnnounceReconcilerUnderTest.processMCSStatus()
-	status := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
-	klog.V(2).InfoS("Received", "actual", status, "expected", expectedStatus)
-	assert.Equal(t, 1, len(status))
-	verifyStatus(t, expectedStatus, status[0])
+	actualStatus := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
+	klog.V(2).InfoS("Received", "actual", actualStatus, "expected", expectedStatus)
+	assert.Equal(t, 1, len(actualStatus))
+	verifyStatus(t, expectedStatus, actualStatus[0])
 }
 
 func TestStatusInNonLeaderCase(t *testing.T) {
@@ -243,10 +255,10 @@ func TestStatusInNonLeaderCase(t *testing.T) {
 	}
 
 	memberClusterAnnounceReconcilerUnderTest.processMCSStatus()
-	status := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
-	klog.V(2).InfoS("Received", "actual", status, "expected", expectedStatus)
-	assert.Equal(t, 1, len(status))
-	verifyStatus(t, expectedStatus, status[0])
+	actualStatus := memberClusterAnnounceReconcilerUnderTest.GetMemberClusterStatuses()
+	klog.V(2).InfoS("Received", "actual", actualStatus, "expected", expectedStatus)
+	assert.Equal(t, 1, len(actualStatus))
+	verifyStatus(t, expectedStatus, actualStatus[0])
 }
 
 func verifyStatus(t *testing.T, expected mcsv1alpha1.ClusterStatus, actual mcsv1alpha1.ClusterStatus) {

--- a/multicluster/controllers/multicluster/mock_membercluster_status_manager.go
+++ b/multicluster/controllers/multicluster/mock_membercluster_status_manager.go
@@ -51,15 +51,15 @@ func (m *MockMemberClusterStatusManager) EXPECT() *MockMemberClusterStatusManage
 }
 
 // AddMember mocks base method.
-func (m *MockMemberClusterStatusManager) AddMember(MemberId common.ClusterID) {
+func (m *MockMemberClusterStatusManager) AddMember(memberID common.ClusterID) {
 	m.ctrl.T.Helper()
-	m.ctrl.Call(m, "AddMember", MemberId)
+	m.ctrl.Call(m, "AddMember", memberID)
 }
 
 // AddMember indicates an expected call of AddMember.
-func (mr *MockMemberClusterStatusManagerMockRecorder) AddMember(MemberId interface{}) *gomock.Call {
+func (mr *MockMemberClusterStatusManagerMockRecorder) AddMember(memberID interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AddMember", reflect.TypeOf((*MockMemberClusterStatusManager)(nil).AddMember), MemberId)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AddMember", reflect.TypeOf((*MockMemberClusterStatusManager)(nil).AddMember), memberID)
 }
 
 // GetMemberClusterStatuses mocks base method.
@@ -77,13 +77,13 @@ func (mr *MockMemberClusterStatusManagerMockRecorder) GetMemberClusterStatuses()
 }
 
 // RemoveMember mocks base method.
-func (m *MockMemberClusterStatusManager) RemoveMember(MemberId common.ClusterID) {
+func (m *MockMemberClusterStatusManager) RemoveMember(memberID common.ClusterID) {
 	m.ctrl.T.Helper()
-	m.ctrl.Call(m, "RemoveMember", MemberId)
+	m.ctrl.Call(m, "RemoveMember", memberID)
 }
 
 // RemoveMember indicates an expected call of RemoveMember.
-func (mr *MockMemberClusterStatusManagerMockRecorder) RemoveMember(MemberId interface{}) *gomock.Call {
+func (mr *MockMemberClusterStatusManagerMockRecorder) RemoveMember(memberID interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "RemoveMember", reflect.TypeOf((*MockMemberClusterStatusManager)(nil).RemoveMember), MemberId)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "RemoveMember", reflect.TypeOf((*MockMemberClusterStatusManager)(nil).RemoveMember), memberID)
 }

